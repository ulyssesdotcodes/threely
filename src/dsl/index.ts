// Main DSL exports - combines all chain types and utilities
import { MockObject3D, mockUtils, mockPresets } from "../three/MockObject3D";

// Object3D chain exports - pure functions (no Node dependencies)
export {
  // Scene management
  setScene,
  getObjectRegistry,
  clearAll,

  // Geometry creation
  sphere,
  box,
  cylinder,
  material,
  mesh,

  // Transforms
  translateX,
  translateY,
  translateZ,
  rotateX,
  rotateY,
  rotateZ,
  applyMock,

  // Rendering
  render,

  // Chain object
  chainObj3d,
} from "./pure-object3d-functions";

// Math chain exports - pure functions (no Node dependencies)
export {
  // Basic operations
  mult,
  add,
  sub,
  div,
  frame,

  // Math functions
  mathAbs,
  mathAcos,
  mathAcosh,
  mathAsin,
  mathAsinh,
  mathAtan,
  mathAtan2,
  mathAtanh,
  mathCbrt,
  mathCeil,
  mathClz32,
  mathCos,
  mathCosh,
  mathExp,
  mathExpm1,
  mathFloor,
  mathFround,
  mathHypot,
  mathImul,
  mathLog,
  mathLog10,
  mathLog1p,
  mathLog2,
  mathMax,
  mathMin,
  mathPow,
  mathRandom,
  mathRound,
  mathSign,
  mathSin,
  mathSinh,
  mathSqrt,
  mathTan,
  mathTanh,
  mathTrunc,

  // Chain object
  chainMath,
} from "./pure-math-functions";

// Parser exports
export {
  parseDSLWithLezer,
  executeDSL,
  logToPanel,
  dslContext,
} from "./parser";

// Converter exports
export { convertASTToNodysseus } from "./direct-ast-to-nodysseus-converter";

// Re-export utilities
export { MockObject3D, mockUtils, mockPresets };

// Individual function imports removed - now only using pure functions without Node dependencies
