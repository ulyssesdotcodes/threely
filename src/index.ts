import { initBackgroundScene } from './three-render';
import { startEditor, setupEditorUI } from './editor';
import { initErrorDisplay } from './error-display';

document.addEventListener('DOMContentLoaded', () => {
  // Initialize error display system first
  initErrorDisplay();
  
  initBackgroundScene();
  startEditor();
  setupEditorUI();
});