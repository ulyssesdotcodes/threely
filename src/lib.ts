// Main library exports for npm publishing
export { parseDSL, executeDSL, dslContext, setScene, getObjectRegistry, clearAll } from './dsl';
// Deprecated: Graph system replaced by direct AST to Nodysseus conversion
// export { Graph, Node, createNode, map, apply } from './graph';
// export type { Node as ThreelyNode } from './graph';
export * as THREE from 'three';