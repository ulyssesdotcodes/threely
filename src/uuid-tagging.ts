// UUID-based function call tagging for CodeMirror and Lezer correlation
import { RangeSet, RangeValue, ChangeDesc, StateField, StateEffect, EditorState, Transaction } from "@codemirror/state";
import { EditorView, ViewPlugin } from "@codemirror/view";
import { parser } from "@lezer/javascript";
import { v4 as uuid } from "uuid";

// RangeValue for UUID-tagged function calls
export class UUIDTag extends RangeValue {
  constructor(public uuid: string, public functionName: string) {
    super();
  }

  eq(other: UUIDTag): boolean {
    return this.uuid === other.uuid && this.functionName === other.functionName;
  }
}

// Store mapping between text positions and UUIDs
export interface FunctionCallInfo {
  uuid: string;
  functionName: string;
  from: number;
  to: number;
  astNodeType: string;
}

// StateEffect for updating the UUID RangeSet
export const setUUIDRangeSet = StateEffect.define<RangeSet<UUIDTag>>();

// StateField for maintaining the UUID RangeSet
export const uuidRangeSetField = StateField.define<RangeSet<UUIDTag>>({
  create() {
    return RangeSet.empty;
  },
  
  update(value, tr) {
    // Check for explicit UUID RangeSet updates
    for (let effect of tr.effects) {
      if (effect.is(setUUIDRangeSet)) {
        return effect.value;
      }
    }
    
    // Map existing ranges through document changes
    if (tr.docChanged) {
      return value.map(tr.changes);
    }
    
    return value;
  }
});

// Global registry for function call UUIDs
const functionCallRegistry = new Map<string, FunctionCallInfo>();

export function getFunctionCallRegistry(): Map<string, FunctionCallInfo> {
  return functionCallRegistry;
}

export function clearFunctionCallRegistry(): void {
  functionCallRegistry.clear();
}

// Generate UUID tags for function calls in the given text
export function generateUUIDTags(text: string): {
  rangeSet: RangeSet<UUIDTag>;
  functionCalls: FunctionCallInfo[];
} {
  const functionCalls: FunctionCallInfo[] = [];
  const ranges: { from: number; to: number; value: UUIDTag }[] = [];
  
  // Parse with Lezer to identify function calls
  const tree = parser.parse(text);
  
  // Walk the tree to find CallExpression nodes
  tree.cursor().iterate((node) => {
    if (node.name === 'CallExpression') {
      const callText = text.slice(node.from, node.to);
      
      // Extract function name from the call expression
      const funcNameMatch = callText.match(/^(\w+)/);
      if (funcNameMatch) {
        const functionName = funcNameMatch[1];
        const callUuid = uuid();
        
        const functionCall: FunctionCallInfo = {
          uuid: callUuid,
          functionName,
          from: node.from,
          to: node.to,
          astNodeType: node.name
        };
        
        functionCalls.push(functionCall);
        functionCallRegistry.set(callUuid, functionCall);
        
        // Create UUID tag for this function call
        const uuidTag = new UUIDTag(callUuid, functionName);
        
        ranges.push({
          from: node.from,
          to: node.to,
          value: uuidTag
        });
      }
    }
  });
  
  // Create RangeSet from the ranges
  const rangeSet = RangeSet.of(ranges.map(r => r.value.range(r.from, r.to)));
  
  return { rangeSet, functionCalls };
}

// Get UUID for a function call at a specific position
export function getUUIDAtPosition(position: number): string | null {
  for (const [uuid, info] of functionCallRegistry) {
    if (position >= info.from && position <= info.to) {
      return uuid;
    }
  }
  return null;
}

// Get UUID from RangeSet at a specific position
export function getUUIDFromRangeSet(rangeSet: RangeSet<UUIDTag>, position: number): string | null {
  let foundUuid: string | null = null;
  
  rangeSet.between(position, position, (from, to, value) => {
    if (position >= from && position <= to) {
      foundUuid = value.uuid;
      return false; // Stop iteration
    }
  });
  
  return foundUuid;
}

// Get UUID from editor state at a specific position
export function getUUIDFromState(state: EditorState, position: number): string | null {
  const rangeSet = state.field(uuidRangeSetField, false);
  if (!rangeSet) return null;
  return getUUIDFromRangeSet(rangeSet, position);
}

// Get function call info by UUID
export function getFunctionCallByUUID(uuid: string): FunctionCallInfo | null {
  return functionCallRegistry.get(uuid) || null;
}

// Get all UUIDs for a given function name
export function getUUIDsForFunction(functionName: string): string[] {
  const uuids: string[] = [];
  for (const [uuid, info] of functionCallRegistry) {
    if (info.functionName === functionName) {
      uuids.push(uuid);
    }
  }
  return uuids;
}

// ViewPlugin to automatically update UUID RangeSet when document changes significantly
export const uuidRangeSetPlugin = ViewPlugin.fromClass(class {
   lastDocText: string = '';
  
  constructor(view: EditorView) {
    this.lastDocText = view.state.doc.toString();
  }
  
  update(update: any) {
    const currentDocText = update.state.doc.toString();
    
    // Only regenerate UUIDs if the document content has changed significantly
    // or if we don't have a UUID RangeSet yet
    if (this.shouldRegenerateUUIDs(update, currentDocText)) {
      this.regenerateUUIDs(update.view, currentDocText);
      this.lastDocText = currentDocText;
    }
  }
  
   shouldRegenerateUUIDs(update: any, currentDocText: string): boolean {
    // Regenerate if:
    // 1. The document changed and text is different
    // 2. There's no existing UUID RangeSet
    // 3. The change was substantial (not just whitespace)
    
    if (!update.docChanged) return false;
    
    const hasExistingRangeSet = update.state.field(uuidRangeSetField, false) && 
                               !update.state.field(uuidRangeSetField, false).size;
    
    if (!hasExistingRangeSet) return true;
    
    // Check if this was a substantial change
    const textChanged = this.lastDocText !== currentDocText;
    const hasNewFunctionCalls = /\w+\s*\(/.test(currentDocText) && textChanged;
    
    return hasNewFunctionCalls;
  }
  
   regenerateUUIDs(view: EditorView, text: string) {
    clearFunctionCallRegistry();
    const { rangeSet } = generateUUIDTags(text);
    
    view.dispatch({
      effects: setUUIDRangeSet.of(rangeSet)
    });
  }
});

// Helper function to create UUID-enabled editor state
export function createUUIDEnabledState(content: string, baseExtensions: any[] = []): EditorState {
  const { rangeSet } = generateUUIDTags(content);
  
  return EditorState.create({
    doc: content,
    extensions: [
      ...baseExtensions,
      uuidRangeSetField,
      uuidRangeSetPlugin,
      setUUIDRangeSet.of(rangeSet)
    ]
  });
}

// Helper function to update existing state with UUID tags
export function updateStateWithUUIDs(state: EditorState): Transaction {
  const content = state.doc.toString();
  clearFunctionCallRegistry();
  const { rangeSet } = generateUUIDTags(content);
  
  return state.update({
    effects: setUUIDRangeSet.of(rangeSet)
  });
}

// CSS for UUID-tagged functions
export const uuidTaggingCSS = `
.uuid-tagged-function {
  position: relative;
}

.uuid-tagged-function::after {
  content: attr(data-function);
  position: absolute;
  top: -20px;
  left: 0;
  background: rgba(0, 0, 0, 0.8);
  color: white;
  padding: 2px 6px;
  border-radius: 3px;
  font-size: 10px;
  white-space: nowrap;
  opacity: 0;
  pointer-events: none;
  transition: opacity 0.2s;
  z-index: 1000;
}

.uuid-tagged-function:hover::after {
  opacity: 1;
}

/* Debug mode - show UUIDs */
.debug-uuids .uuid-tagged-function::after {
  content: attr(data-uuid);
  opacity: 0.7;
}
`;