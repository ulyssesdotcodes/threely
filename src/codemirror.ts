import { EditorState, Prec, Compartment } from "@codemirror/state";
import { EditorView, keymap } from "@codemirror/view";
import { basicSetup } from "codemirror";
import { javascript } from "@codemirror/lang-javascript";
import { vim } from "@replit/codemirror-vim";
import { getTextBlockAtPosition } from "./text_utils";
import { executeDSL } from "./dsl";
import {
  uuidRangeSetField,
  uuidRangeSetPlugin,
  generateUUIDTags,
  setUUIDRangeSet,
  getUUIDFromState,
  UUIDTag,
} from "./uuid-tagging";
import { State } from "./nodysseus/node-types";

export { EditorState, EditorView, keymap, basicSetup, javascript };

// Vim mode state management
const VIM_MODE_KEY = "three-tree-vim-mode";
const vimCompartment = new Compartment();
let currentEditorView: EditorView | null = null;

export function getVimModeEnabled(): boolean {
  try {
    return localStorage.getItem(VIM_MODE_KEY) === "true";
  } catch {
    return false;
  }
}

export function setVimModeEnabled(enabled: boolean): void {
  try {
    localStorage.setItem(VIM_MODE_KEY, enabled.toString());
    if (currentEditorView) {
      currentEditorView.dispatch({
        effects: vimCompartment.reconfigure(enabled ? vim() : []),
      });
    }
  } catch {
    // Ignore localStorage errors
  }
}

export function getCurrentEditorView(): EditorView | null {
  return currentEditorView;
}

// Custom command for Ctrl-Enter
const handleCtrlEnter = (view: EditorView): boolean => {
  const blockInfo = getBlockAtCursor(view);

  if (blockInfo && blockInfo.block) {
    const code = blockInfo.block.trim();

    try {
      // const ranges: { start: number; end: number; uuid: UUIDTag }[] = [];
      // .between(
      //   blockInfo.start,
      //   blockInfo.end,
      //   (start, end, uuid) => (
      //     ranges.push({
      //       start: start - blockInfo.start,
      //       end: end - blockInfo.start,
      //       uuid,
      //     }),
      //     undefined
      //   ),
      // );
      const result = executeDSL(
        code,
        view.state.field(uuidRangeSetField),
        undefined,
        blockInfo.start,
      );
      if (result) {
      } else {
      }
    } catch (error) {
      console.error("Error executing DSL code:", error);
    }
  }

  // Prevent the default new line behavior by returning true
  return true;
};

// Function to get block at cursor position
export const getBlockAtCursor = (
  view: EditorView,
): { block: string; start: number; end: number } | null => {
  const text = view.state.doc.toString();
  const cursorPos = view.state.selection.ranges[0].from;

  return getTextBlockAtPosition(text, cursorPos);
};

export const defaultContent = `mesh(sphere(), material()).translateX(1).rotateY(45).render("mySphere")

mesh(sphere(), material()).translateX(frame().mult(0.02).floor()).rotateY(45).render("mySphere")

// Try pressing Ctrl+Enter on the line above!
// This will create a sphere mesh named "mySphere", translate it, rotate it, and add it to the scene
// Running it again will update the existing object instead of creating a new one!
// Note: sphere(), box(), and cylinder() now create mock geometries that are applied during render!

mesh(box(2, 1, 1), material({color: 0xff0000})).translateX(-3).render("redBox")

mesh(cylinder(), material({color: 0x0000ff, wireframe: true})).translateX(3).render("blueCylinder")

// Animated sphere using frame counter:
mesh(sphere(0.5), material({color: 0x00ff00})).translateX(Math.sin(frame() * 0.01) * 2).render("animatedSphere")

// Using mock objects for quick property setting:
applyMock(mesh(sphere(), material({color: 0xff0000})), mockUtils.position(2, 1, 0)).render("mockSphere")

// Using presets for common configurations:
applyMock(mesh(box(), material()), mockPresets.elevated(3)).render("elevatedBox")

// Custom geometry parameters with mock system:
mesh(sphere(1.5, 16, 8), material({color: 0xff00ff})).render("customSphere")

// Custom box with specific dimensions:
mesh(box(3, 0.5, 0.5), material({color: 0x00ffff})).translateY(-2).render("customBox")

// Custom cylinder with different top/bottom radius:
mesh(cylinder(0.5, 1, 2), material({color: 0xffff00})).translateZ(-3).render("customCylinder")

// Try modifying the values and re-running to see objects update:
// mesh(sphere(0.5), material({color: 0xffffff})).translateY(2).rotateX(90).render("mySphere")
// mesh(box(1, 3, 1), material({color: 0x00ff00})).translateX(-2).render("redBox")

// Use clearAll() to remove all objects:
// clearAll()
`;

export function createEditorState(
  content: string = defaultContent,
): EditorState {
  const vimModeEnabled = getVimModeEnabled();

  // Generate new tags for the content
  const { rangeSet } = generateUUIDTags(content);

  const state = EditorState.create({
    doc: content,
    extensions: [
      vimCompartment.of(vimModeEnabled ? vim() : []),
      basicSetup,
      javascript(),
      Prec.highest(keymap.of([{ key: "Ctrl-Enter", run: handleCtrlEnter }])),
      uuidRangeSetField, // Add UUID RangeSet field
      uuidRangeSetPlugin, // Add UUID RangeSet plugin for automatic updates
    ],
  });

  // Apply the initial UUID RangeSet
  return state.update({
    effects: setUUIDRangeSet.of(rangeSet),
  }).state;
}

export function setCurrentEditorView(view: EditorView): void {
  currentEditorView = view;
}
