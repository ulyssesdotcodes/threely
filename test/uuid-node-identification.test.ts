// Tests for UUID-based node identification system
import { convertASTToNodysseus } from "../src/dsl/direct-ast-to-nodysseus-converter";
import { generateUUIDTags } from "../src/uuid-tagging";

describe("UUID-based Node Identification", () => {
  it("should assign UUIDs to all nodes when available", () => {
    const code = "sphere(1)";
    const { rangeSet } = generateUUIDTags(code);

    const result = convertASTToNodysseus(code, rangeSet);

    // Every node should have a UUID
    Object.values(result.graph.nodes).forEach((node) => {
      expect(node.id).toBeDefined();
      expect(typeof node.id).toBe("string");
      expect(node.id.length).toBeGreaterThan(0);
    });
  });

  it("should generate fallback IDs when no UUID range found", () => {
    const code = "sphere(1)";
    // Use empty range set to force fallback ID generation
    const emptyRangeSet = generateUUIDTags("").rangeSet;

    const result = convertASTToNodysseus(code, emptyRangeSet);

    // Should still have node IDs (fallback)
    Object.values(result.graph.nodes).forEach((node) => {
      expect(node.id).toBeDefined();
      expect(typeof node.id).toBe("string");
      // Fallback IDs should start with "fallback-"
      if (node.id.startsWith("fallback-")) {
        expect(node.id).toMatch(/^fallback-\d+-\d+-/);
      }
    });
  });

  it("should preserve UUID metadata in nodes", () => {
    const code = "sphere(1)";
    const { rangeSet } = generateUUIDTags(code);

    const result = convertASTToNodysseus(code, rangeSet);

    // At least some nodes should have UUID metadata
    const nodesWithUUIDs = Object.values(result.graph.nodes).filter(
      (node) => "uuid" in node && node.uuid,
    );

    expect(nodesWithUUIDs.length).toBeGreaterThan(0);
  });

  it("should use consistent IDs for the same AST positions", () => {
    const code = "sphere(1)";
    const { rangeSet } = generateUUIDTags(code);

    const result1 = convertASTToNodysseus(code, rangeSet);
    const result2 = convertASTToNodysseus(code, rangeSet);

    // Node IDs should be consistent between conversions
    const nodeIds1 = Object.keys(result1.graph.nodes).sort();
    const nodeIds2 = Object.keys(result2.graph.nodes).sort();

    expect(nodeIds1).toEqual(nodeIds2);
  });

  // Snapshot tests for UUID generation
  describe("UUID Generation Snapshots", () => {
    it("should generate consistent UUIDs for simple function calls", () => {
      const code = "sphere(1)";
      const { rangeSet, functionCalls } = generateUUIDTags(code);

      expect(functionCalls).toMatchSnapshot();
    });

    it("should generate consistent UUIDs for math expressions", () => {
      const code = "Math.sin(3.14 / 2)";
      const { rangeSet, functionCalls } = generateUUIDTags(code);

      expect(functionCalls).toMatchSnapshot();
    });

    it("should generate consistent UUIDs for object expressions", () => {
      const code = "{ x: 1, y: 2, z: 3 }";
      const { rangeSet, functionCalls } = generateUUIDTags(code);

      expect(functionCalls).toMatchSnapshot();
    });

    it("should generate consistent UUIDs for chained method calls", () => {
      const code = "sphere(1).translateX(2).rotateY(0.5)";
      const { rangeSet, functionCalls } = generateUUIDTags(code);

      expect(functionCalls).toMatchSnapshot();
    });
  });

  // Snapshot tests for AST to Nodysseus conversion
  describe("AST to Nodysseus Conversion Snapshots", () => {
    it("should convert simple function calls consistently", () => {
      const code = "sphere(1)";
      const { rangeSet } = generateUUIDTags(code);
      const result = convertASTToNodysseus(code, rangeSet);

      expect(result.graph).toMatchSnapshot();
    });

    it("should convert math expressions consistently", () => {
      const code = "Math.sin(3.14 / 2)";
      const { rangeSet } = generateUUIDTags(code);
      const result = convertASTToNodysseus(code, rangeSet);

      expect(result.graph).toMatchSnapshot();
    });

    it("should convert object expressions consistently", () => {
      const code = "{ x: 1, y: 2, z: 3 }";
      const { rangeSet } = generateUUIDTags(code);
      const result = convertASTToNodysseus(code, rangeSet);

      expect(result.graph).toMatchSnapshot();
    });

    it("should convert chained method calls consistently", () => {
      const code = "sphere(1).translateX(2).rotateY(0.5)";
      const { rangeSet } = generateUUIDTags(code);
      const result = convertASTToNodysseus(code, rangeSet);

      expect(result.graph).toMatchSnapshot();
    });

    it("should convert complex nested expressions consistently", () => {
      const code =
        "box({ x: Math.sin(frame), y: 2, z: 3 }).translateX(frame * 0.1)";
      const { rangeSet } = generateUUIDTags(code);
      const result = convertASTToNodysseus(code, rangeSet);

      expect(result.graph).toMatchSnapshot();
    });
  });
});
