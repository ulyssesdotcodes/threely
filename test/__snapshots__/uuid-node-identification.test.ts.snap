// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UUID-based Node Identification AST to Nodysseus Conversion Snapshots should convert chained method calls consistently 1`] = `
{
  "edges": {},
  "edges_in": {},
  "id": "direct-nodysseus-graph-0fe9d32d-f149-4cba-a31c-0e41b77ec485",
  "nodes": {
    "0fe9d32d-f149-4cba-a31c-0e41b77ec485": {
      "id": "0fe9d32d-f149-4cba-a31c-0e41b77ec485",
      "ref": "@error.unknown_function",
      "uuid": "0fe9d32d-f149-4cba-a31c-0e41b77ec485",
      "value": "rotateY",
    },
    "8ed075bb-28a0-4df9-9248-aca3784617bb": {
      "id": "8ed075bb-28a0-4df9-9248-aca3784617bb",
      "ref": "@error.unknown_function",
      "uuid": "8ed075bb-28a0-4df9-9248-aca3784617bb",
      "value": "sphere",
    },
    "e5a3b2be-f387-42c2-ae5e-d4bce1daf920": {
      "id": "e5a3b2be-f387-42c2-ae5e-d4bce1daf920",
      "ref": "@error.unknown_function",
      "uuid": "e5a3b2be-f387-42c2-ae5e-d4bce1daf920",
      "value": "translateX",
    },
  },
  "out": "0fe9d32d-f149-4cba-a31c-0e41b77ec485",
}
`;

exports[`UUID-based Node Identification AST to Nodysseus Conversion Snapshots should convert complex nested expressions consistently 1`] = `
{
  "edges": {},
  "edges_in": {},
  "id": "direct-nodysseus-graph-83650e11-14e1-4d00-9e05-5d9931aeeae0",
  "nodes": {
    "4eab8308-6f3f-4243-a66d-412a03745b94": {
      "id": "4eab8308-6f3f-4243-a66d-412a03745b94",
      "ref": "@error.unknown_function",
      "uuid": "4eab8308-6f3f-4243-a66d-412a03745b94",
      "value": "box",
    },
    "83650e11-14e1-4d00-9e05-5d9931aeeae0": {
      "id": "83650e11-14e1-4d00-9e05-5d9931aeeae0",
      "ref": "@error.unknown_function",
      "uuid": "83650e11-14e1-4d00-9e05-5d9931aeeae0",
      "value": "translateX",
    },
  },
  "out": "83650e11-14e1-4d00-9e05-5d9931aeeae0",
}
`;

exports[`UUID-based Node Identification AST to Nodysseus Conversion Snapshots should convert math expressions consistently 1`] = `
{
  "edges": {},
  "edges_in": {},
  "id": "direct-nodysseus-graph-ddee4e47-cfd8-45f2-8641-9e12304eb1da",
  "nodes": {
    "ddee4e47-cfd8-45f2-8641-9e12304eb1da": {
      "id": "ddee4e47-cfd8-45f2-8641-9e12304eb1da",
      "ref": "@error.unknown_function",
      "uuid": "ddee4e47-cfd8-45f2-8641-9e12304eb1da",
      "value": "sin",
    },
  },
  "out": "ddee4e47-cfd8-45f2-8641-9e12304eb1da",
}
`;

exports[`UUID-based Node Identification AST to Nodysseus Conversion Snapshots should convert object expressions consistently 1`] = `
{
  "edges": {},
  "edges_in": {},
  "id": "direct-nodysseus-graph-fallback-0-1-_",
  "nodes": {
    "fallback-0-1-_": {
      "id": "fallback-0-1-_",
      "value": null,
    },
  },
  "out": "fallback-0-1-_",
}
`;

exports[`UUID-based Node Identification AST to Nodysseus Conversion Snapshots should convert simple function calls consistently 1`] = `
{
  "edges": {},
  "edges_in": {},
  "id": "direct-nodysseus-graph-83e9d78b-9034-4e02-bf6a-b8c168bbe4c8",
  "nodes": {
    "83e9d78b-9034-4e02-bf6a-b8c168bbe4c8": {
      "id": "83e9d78b-9034-4e02-bf6a-b8c168bbe4c8",
      "ref": "@error.unknown_function",
      "uuid": "83e9d78b-9034-4e02-bf6a-b8c168bbe4c8",
      "value": "sphere",
    },
  },
  "out": "83e9d78b-9034-4e02-bf6a-b8c168bbe4c8",
}
`;

exports[`UUID-based Node Identification UUID Generation Snapshots should generate consistent UUIDs for chained method calls 1`] = `
[
  {
    "astNodeType": "CallExpression",
    "from": 24,
    "functionName": "rotateY",
    "to": 31,
    "uuid": "8576ea16-b7a8-4fc8-abf3-141687cd4634",
  },
  {
    "astNodeType": "CallExpression",
    "from": 10,
    "functionName": "translateX",
    "to": 20,
    "uuid": "86691f96-6d74-4d3d-accf-3ce2645bda2d",
  },
  {
    "astNodeType": "CallExpression",
    "from": 0,
    "functionName": "sphere",
    "to": 6,
    "uuid": "febaf0bc-9c8d-4ad7-b6d9-a62e565b8246",
  },
  {
    "astNodeType": "Number",
    "from": 7,
    "functionName": "1",
    "to": 8,
    "uuid": "e943f582-07ec-4a76-8ccc-ec32e6435adc",
  },
  {
    "astNodeType": "Number",
    "from": 21,
    "functionName": "2",
    "to": 22,
    "uuid": "3126aaf1-b5ff-4453-a76e-d0d2e0edf914",
  },
  {
    "astNodeType": "Number",
    "from": 32,
    "functionName": "0.5",
    "to": 35,
    "uuid": "f6e56f8c-0b20-4791-afd4-f20ef7f3f03d",
  },
]
`;

exports[`UUID-based Node Identification UUID Generation Snapshots should generate consistent UUIDs for math expressions 1`] = `
[
  {
    "astNodeType": "CallExpression",
    "from": 5,
    "functionName": "sin",
    "to": 8,
    "uuid": "56c8e7e3-ecd3-4e54-b656-c392824ffa06",
  },
  {
    "astNodeType": "Number",
    "from": 9,
    "functionName": "3.14",
    "to": 13,
    "uuid": "33444940-f645-4224-9a9a-be7d1e748f9d",
  },
  {
    "astNodeType": "Number",
    "from": 16,
    "functionName": "2",
    "to": 17,
    "uuid": "e8a23617-0968-47bd-a391-7424bcc9332f",
  },
]
`;

exports[`UUID-based Node Identification UUID Generation Snapshots should generate consistent UUIDs for object expressions 1`] = `
[
  {
    "astNodeType": "Number",
    "from": 5,
    "functionName": "1",
    "to": 6,
    "uuid": "1622a1ea-28e5-43ad-969f-d9982d97d625",
  },
  {
    "astNodeType": "Number",
    "from": 11,
    "functionName": "2",
    "to": 12,
    "uuid": "a5dc5f0c-e094-465e-809f-c05446fa6bc1",
  },
  {
    "astNodeType": "Number",
    "from": 17,
    "functionName": "3",
    "to": 18,
    "uuid": "0a775fc5-b568-482c-a6b8-7e637ffbe9c6",
  },
]
`;

exports[`UUID-based Node Identification UUID Generation Snapshots should generate consistent UUIDs for simple function calls 1`] = `
[
  {
    "astNodeType": "CallExpression",
    "from": 0,
    "functionName": "sphere",
    "to": 6,
    "uuid": "3c009e9f-78a1-4378-be19-d331fa9a1060",
  },
  {
    "astNodeType": "Number",
    "from": 7,
    "functionName": "1",
    "to": 8,
    "uuid": "70fd8ea9-c79c-49bd-8cfc-0ed90cf0215e",
  },
]
`;
